README.rst
Python scripts to plot PECUBE modeling results
==============================================

Xavier Robert, Grenoble,                         

This tool have been set up to help the plot of PECUBE (Braun, 2003; Braun et al., 2012) modeling results.

    - The Plot-PecubeV4+-Forward.py permits to plot predictions versus observations for Pecube V4+ forward modeling.

    - The Inversion_plots-XR.py permits to plot the results of NA Inversions (Sambridge, 1999 a & b) with scatter plots and 1D / 2D marginals. To plot the marginals, the results of NA Inversions should be analyzed with NAB.

Usage
-----
	
After copying the scripts in your folder ant editing the main sections to fit with your project, each script can be ran in a terminal with the command : 

.. code-block:: bash

	~$ python pyPlotPecubeForward.py

or:

.. code-block:: bash

    ~$ python pyPlotPecubeNA.py

Please, edit the Python files in your text-editor and read the headers of the main section in each files to know how to parametrize them. 

The second solution to install the module with pip:

.. code-block:: bash

    ~$ pip install pyPlotPecube

And then, inside a Python environement, import the function you need:

.. code-block:: python

    >>> from pyPlotPecube import pyPlotPecubeForward as Fplot

    >>> from pyPlotPecube import pyPlotPecubeNA as NAplot

and finally run the function in your Python environement :

.. code-block:: python

    >>> Fplot.PlotPecubeForward(dataplot = dataplot, graphpath = graphpath, datafnme = datafnme, inputdata = inputdata, graphtitle = graphtitle, agerange = agerange, profiletype = profiletype, size_x = size_x, size_y = size_y, A = A, B = B)

    >>> NAplot.NAplot_Pecube(param = param, dataplot = dataplot, inv_results = inv_results, data_nab = data_nab, graph_path  = graph_path, PDF_1D = PDF_1D, PDF_2D = PDF_2D, pdf1d_results = pdf1d_results, pdf2d_results = pdf2d_results, tick_space = tick_space, size_x = size_x, size_y = size_y, size_m = size_m, size_mis = size_mis)

You can also write a script that facilitate the call of the function with the right input data. See Templates in Tests/.

Parameters
----------
.. .. currentmodule:: pyPlotPecube/pyPlotPecubeForward

.. .. autofunction:: PlotPecubeForward

PlotPecubeForward
^^^^^^^^^^^^^^^^^

For Forward plots, arguments are :

    1. ``datafnme`` (string): results of Pecube forward modeling.

        No default value.

	2. ``inputdata`` (string): input data declared in Pecube.in. This is used to plot the errorbars.

        No default value.

    3. ``inputdataPTt`` (str, optional): Path and file name of the input PTt data file. Need to be given if ``'TTp'`` in ``dataplot``.
            
        Default = ``None`` .

    4. ``outputdataPTt`` (str, optional): Path and file name of the output PTt prediction/comparison file. Usually, this is ``'output/CompareTT.csv'``. Need to be given if ``'TTp'`` in ``dataplot``.
   
         Default = ``None``.
    
    5. ``stackTT`` (boolean, optional): Tells if we want to plot all the PTt path in one single graph
									 
        Defaults = ``True``.
    
    6. ``inputdataFTL`` (str, optional): Path and file name of the input FTL data file. Need to be given if ``'FTL'`` in ``dataplot``.
        
        Default = ``None``.

    7. ``outputdataFTL`` (str, optional): Path and file name of the output FTL prediction/comparison file. Usually, this is ``'output/CompareFTL.csv'`` Need to be given if ``'FTL'`` in ``dataplot``.
    
        Default = ``None``.

	8. ``dataplot`` (list, optional): List of data to plot: ``['AHe', 'AFT', 'ZHe', 'ZFT', 'KAr', 'MAr', 'BAr', 'MTL', 'TTp']``; by default, the altitude will be plotted; Do not forget the simple quotes !!! 
        
        Note: For the moment, ``MTL`` is not implemented.

        If ``'TTp'``, you should give the PTT files names (``inputdataPTt`` and ``outputdataPTt``)
        
        Defaults = ``['AHe','AFT']``.

	9. ``graphpath`` (str, optional): name of the folder where the plot will be written. Usually you do not have to change it.
        
        Defaults = ``'Graphs'``.

	10. ``graphtitle`` (str, optional): title to write on the graph. 
        
        Defaults = ``None``.

	11. ``agerange`` (2*1 array of floats, optional): range of the ages to plot on the profiles ``[min, max]``.
        
        Defaults = ``None``.

	12. ``profiletype`` (list, optional): type of profile, could be one or more of ``['Latitude', 'Longitude', 'Altitude', 'Projected']``. If ``[]``, no age profile is plotted.
        
        Defaults = ``[]``.

    13. ``size_x`` (int, optional): Font size for x-axis. 
    
        Defaults = ``15``.
        
	14. size_y (int, optional): Font size for y-axis. 
        
        Defaults = ``15``.

	15. ``A``, ``B`` (floats, optional): If need of a projected transect, define the line along which we will project with the coordinate of the point A and B defining, respectivelly, the begining and the end of the transect, in lat-long/WGS84.
        
        Defaults = ``None``.
    
    16. ``agename`` (dict, optional): legend of each data system         
		
        If ``None``, this is set to
        
            {'AHe' : 'AHe (Ma)',
		
            'AFT' : 'AFT (Ma)',
		
            'ZHe' : 'ZHe (Ma)',
			
            'ZFT' : 'ZFT (Ma)',
			
            'KAr' : 'KAr (Ma)',
			
            'BAr' : 'Biot. Ar (Ma)',
			
            'MAr' : 'Musc. Ar (Ma)',
			
            'HbAr' : 'Hb Ar (Ma)',
			
            'FTL' : 'FT length (µm)'}
			
        Default = ``None``. 
	
    17. ``predname`` (dict, optional): legend of each predicted system.
        
        If ``None``, this is set to:
            
            {'AHe' : 'Predicted AHe (Ma)',
            
            'AFT' : 'Predicted AFT (Ma)',
            
            'ZHe' : 'Predicted ZHe (Ma)',
            
            'ZFT' : 'Predicted ZFT (Ma)',
            
            'KAr' : 'Predicted KAr (Ma)',
            
            'BAr' : 'Predicted Biot. Ar (Ma)',
            
            'MAr' : 'Predicted Musc. Ar (Ma)',
            
            'HbAr' : 'Predicted Hb Ar (Ma)',
            
            'FTL' : 'Predicted FT length (µm)'}         
        
        Default = ``None``. 
	
    18. ``colores`` (dict, optional): Colors used for the different age system
    
        If ``None``, this is set to:
            
            {'AHe' : 'y',
            
            'AFT' : 'r',
            
            'ZHe' : 'g',
            
            'ZFT' : 'b',
            
            'KAr' : 'k',
            
            'BAr' : 'c',
            
            'MAr' : 'm',
            
            'HbAr' : '0.75',
            
            'FTL' : 'y'}
        
        Default = ``None``.

NAplot_Pecube
^^^^^^^^^^^^^

For Inverse plots, arguments are :

    1. ``param`` (list of str): Define as many variable as you have, with their unit. Check the order in ``na.sum`` (open it with a text editor) or in ``NA_Results``. If you use the later, first column is always the misfit. /!\ If you want to use a ``slash`,  use this one --> ``'∕'`` <-- . The regular slash it is a unicode division symbol windows and OSX do not allow the use of it.
        
        Exemple : ``param = ['Offset (km)','Basal Temperature (°C)','Slip rate (km∕Ma)']``
                             
    
    2. ``dataplot`` (list of couple of integers): Set the couple of variables to plot against each other. If you plot 2D pdfs (contours), please, CHECK that the couple of parameters to plot are the same and in the same order than in the nab.in file. This python script checks it and will insult you if this is not compatible !!!
        
        Exemple : Offset vs Slip rates, ``dataplot = [(1,3)]`` -->  plot=(1,3); 
        
        No default value.
    
    3. ``tick_space`` (array of floats): Set the space between ticks for x and y axes for each parameters (same order than the list param). If the tick format does not fit your variables, you may need to modify the dictionnary ``tick_order`` in the function multiplot.

        No default value.
    
    4. ``inv_results`` (str, optional): Name of the NA file with the inversion results, usually ``NA_Results``.
        
        Defaults = ``'NA/NA_results.csv'``.
    
    5. ``data_nab`` (str, optional): Name of the NAB file with the inversion results, usually ``nab.out``. 
        
        Defaults = ``'NA/NAB/nab.out'``.
    
    6. ``graph_path`` (str, optional) : Path where to save graphs and results Usually NA/Graphs. 
        
        Do not forget the ``'/'`` at the end. 
        
        Defaults = ``'NA/Graphs/'``.
    
    7. ``PDF_1D`` (bool, optional): Choose if you want the 1-PDFs (Probability Density Function); 
        
        Defaults to ``True``.
    
    8. ``PDF_2D`` (bool, optional): Choose if you want the 2-PDFs (Probability Density Function); 
        
        Defaults = ``False``.
    
    9. ``pdf1d_results`` (string, optional): Print the 1-pdfs in a text file. 
        
        Defaults = ``None``.
    
    10. ``pdf2d_results`` (string, optional): Print the 2-pdfs in a text file. 

        Defaults = ``None``.
    
    11. ``size_x`` (int, optional): Size of the font for the x axes label. 

        Defaults = ``15``.
    
    12. ``size_y`` (int, optional): Size of the font for the y axes label. 
        
        Defaults = ``15``.
    
    12. ``size_m`` (int, optional): Size of the font for the markers label. 
    
        Defaults = ``15``.
    
    13. ``size_mis`` (int, optional): Size of the markers of the misfits. 
        
        Defaults = ``50``.
    
    14. ``peak_thres`` (float, optional): Threshold to find peaks; between 0. and 1. See peakutils documentation.
                                      
        Default = ``0.05``.

    15. peak_min_dist (interger, optional): Minimum distance between the peaks. See peakutils documentation.
        
        Default = ``30``.

Examples
--------

Two example's sripts are in the ``Tests/`` folder. Just run them from their location :

.. code-block:: bash
    
    ~$ python Template_ForwardPlot.py
    
    ~$ python Template_NAPlot.py

For instance, the Template_ForwardPlot.py script gives the plot:

.. figure:: Tests/Graphs/Forward.png
   :scale: 50 %
   :align: center
   :figclass: align-center 

   **Figure 1:** Data and predictions along a projected transect.

and the Template_NAPlot.py script permits to build the plots:

.. figure:: Tests/Graphs/NA-1.png
   :scale: 50 %
   :align: center
   :figclass: align-center 


History.txt
History changes of the python module pyPlotPecube

1.3.1 (2021/10/04)
	- Add stacked PTt path plot
	- Ameliorate the output structure
	- Prepare for MTL plots
	- Colors in terminal output

1.3.0 (2021/09/29)
	- Add plot of input and predicted PTt paths
	- Update Readme accordingly
	- Add possibility to change Casse in the input files
	- Clean/simplify the code

1.2.3 (2021/09/24)
	- Add age and predictions comparison graph, with errorbars

1.2.2 (2021/09/21):
	- Clean plot forward
	- Clean plot NA
	- Correct Readme 

1.2.1 (2021/09/22):
	- Add peakutils parameters as inputs
	- Add colors in the terminal
	- Correct Runtime Error bug with curve_fit
	- Remove non acceptable parts of the graphs

1.2.0 (2021/09/22): 
	- Find number of peaks, 
	- Compute the stats for each peak
	- Built the 2 functions (Forward and NA) as one single python module

1.1.0 (2021/09/20):
	- Fit 1D NA results with a 1-peak gaussian, 
	- Plot the fit
    - Compute the stats (Mean & 1 sigma error)

1.0.1: 2021/02/18
	- Initial preliminary release for NA plots

0.2.0: 2021/07/21 
	- First work with Pecube V4+ forward version
	- Add projection for projected transects

0.1.0: 2015/06/01 
	- First work with old Pecube forward version
	- GPL 3.0 Licence


Plot-Pecube-ForwardV3.py
######!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Script to plot Pecube results in forward mode
By Xavier Robert
Grenoble, 2015.06.01

USAGE :
  1- Copy this file in the main Pecube/ folder where run.sh is stored
  2- Add the line "python Plot-Pecube-Forward.py at the end of the file run.sh
  3- Set which system you want to plot in the "# Define data to analysis" section (see below)
  4- Run ./run.sh
  2to4bis - Run in the terminal: $ python Plot-Pecube-Forward.py after the run of ./bin/Pecube
  5- The plots will be in the folder RUNXX/Graphs/ where RUNXX/ is the working folder
               set in topo_parameter.txt

INPUTS:
The inputs are in the script file, in the "# Define data to analysis" section. 
The different arguments are described.

xavier.robert@ujf-grenoble.fr

(c) licence CCby-nc : http://creativecommons.org/licenses/by-nc/3.0/ 2015

"""
###### History :  #######
#    - ??? : First release for Pecube V2 and V3
#    - 2021//: Update for Pecube V4
###### End History #######

###### To DO :  #######
#	 - Adapt the script for Pecube version 4.x (not same input file !)
#    - Test the MTL pdfs (I have not tested it, even if it should be OK)
#    - Add error bars on data graphs ? 
#      For that we should read the input data file given in topo_parameters.txt 
#                    (paragraphe N°12 of topo_parameters.txt)
#    - Add the plot of a projected profile between point A and point B ?
#   - 
#      
###### End To DO #######

from __future__ import  division
# This to be sure that the result of the division of integers is a real, not an integer

# Import modules
import sys
import os
import copy
import numpy as np
import scipy as sp
import matplotlib.pyplot as plt
import shutil


#######################################
# Define Pecube settings
# VPecube : Version of Pecube used; Can be 2, 3 or 4
#VPecube = 2
#VPecube = 3
VPecube = 4
# pfolder is the folder of the Pecube Run to plot
pfolder = RUN00




# Define data to analysis
# dataplot: List of data to plot ; By default, the altitude will be plotted
#           Do not forget the simple quotes !!!
#           ['AHe', 'AFT', 'ZHe', 'ZFT', 'KAr', 'MAr', 'BAr', 'MTL']
dataplot = ['AFT', 'ZFT']




# graphpath: name of the folder where the plot will be written
#            Usually you do not have to change it
graphpath = 'Graphs'

# profiletype: type of profile = ['Latitude', 'Longitude', 'Altitude']
#              if [], no age profile is plotted
#              For the moment, NO there is no projected profile; This is one thing to add ?
profiletype = ['Latitude', 'Longitude', 'Altitude']
#profiletype = ['Latitude', 'Altitude']

# end define the data and parameters
#######################################


# Dictionaries definitions:
# You may have to change it depending on your Pecube version, and on your settings...
# agecol: respective column number of the data in the file comparison.txt
#         needs to be changed depending on your settings in the last paragraph 
#         of the file topo_parameters.txt
agecol = {'alt' : 3,
          'AHe' : 5,
          'AFT' : 7,
          'ZHe' : 9,
          'ZFT' : 11,
          'KAr' : 13,
          'BAr' : 15,
          'MAr' : 17,
          'HbAr' : 19,
          'FTL' : 21
          }
# errcol: respective column number of the error on data in the data input file
errcol = {'AHe' : 5,
          'AFT' : 7,
          'ZHe' : 9,
          'ZFT' : 11,
          'KAr' : 13,
          'BAr' : 15,
          'MAr' : 17,
          'HbAr' : 19,
          'FTL' : 21
          }
# agename: legend of each data system         
agename = {'AHe' : 'AHe (Ma)',
           'AFT' : 'AFT (Ma)',
           'ZHe' : 'ZHe (Ma)',
           'ZFT' : 'ZFT (Ma)',
           'KAr' : 'KAr (Ma)',
           'BAr' : 'Biot. Ar (Ma)',
           'MAr' : 'Musc. Ar (Ma)',
           'HbAr' : 'Hb Ar (Ma)',
           'FTL' : 'FT length (µm)'
           }
# predname: legend of each predicted system         
predname = {'AHe' : 'Predicted AHe (Ma)',
            'AFT' : 'Predicted AFT (Ma)',
            'ZHe' : 'Predicted ZHe (Ma)',
            'ZFT' : 'Predicted ZFT (Ma)',
            'KAr' : 'Predicted KAr (Ma)',
            'BAr' : 'Predicted Biot. Ar (Ma)',
            'MAr' : 'Predicted Musc. Ar (Ma)',
            'HbAr' : 'Predicted Hb Ar (Ma)',
            'FTL' : 'Predicted FT length (µm)'
           }
# plotprof: respective column number of the data in the file comparison.txt
plotprof = {'Latitude' : 1, 
            'Longitude' : 0, 
            'Altitude' : 2
            }
# colores: Colors used for the different age system
colores = {'AHe' : 'y',
           'AFT' : 'r',
           'ZHe' : 'g',
           'ZFT' : 'b',
           'KAr' : 'k',
           'BAr' : 'c',
           'MAr' : 'm',
           'HbAr' : '0.75',
           'FTL' : 'y'
           }

# Tags: Tags used in Pecube v4+ input files
#tagsName = ('SAMPLE', 'LON', 'LAT', 'HEIGHT', 
tagsName = ('AHE', 'DAHE', 
		    'AFT', 'DAFT',
		    'ZHE', 'DZHE',
		    'ZFT', 'DZFT',
		    'KAR', 'DKAR',
		    'BAR', 'DBAR',
		    'MAR', 'DMAR',
		    'HAR', 'DHAR')

# tags: Tags used in Pecube v4+ input files
#tags = ('SAMPLE', 'LON', 'LAT', 'HEIGHT', 
tags = {'HEIGHT' : 'alt', 
        'AHE'    : 'AHe',
		'DAHE'   : 'AHe', 
	    'AFT'    : 'AFT',
		'DAFT'   : 'AFT',
		'ZHE'    : 'ZHe',
		'DZHE'   : 'ZHe',
	    'ZFT'    : 'ZFT',
		'DZFT'   : 'ZFT',
	    'KAR'    : 'KAr',
		'DKAR'   : 'KAr',
		'BAR'    : 'BAr',
		'DBAR'   : 'BAr',
	    'MAR'    : 'MAr',
		'DMAR'   : 'MAr',
	    'HAR'    : 'HAr', 
		'DHAR'   : 'HAr'
		}


tagsMTL = ('SIZE','FTL')

tagsQT = ('TIMEH', 'TEMPH', 'DTEMPH',
		  'AGE43', 'DAGE43', 'DUR43', 'TEMP43', '%REL', 'D%REL', 'AREL', 'DAREL')

tagsTL = ('DOSER', 'D0', 'ATL', 'BTL', 'ET', 'LOGS', 'LOGRHO', 'N/N')

############################################################################
######## Beginning of functions definition ###########
def plottot(datac, i, xlabel, ylabel, err = None):
	"""
	Function to plot the data

	INPUT:
		datac  :
		i      :
		xlabel :
		ylabel :
		err    : ; default = None
	OUTPUTS:
		No outputs

	USAGE:
		plottot(datac, i, xlabel, ylabel, err)
		or
		plottot(datac, i, xlabel, ylabel)

	"""
	if err == None:
		plt.plot(datac[:, i-1], datac[:, i], 'o', color = 'r')
	else:
		plt.errorbar(datac[:, i-1], datac[:, i], xerr = err, fmt = 'o', color = 'r')
		
	# Find axes limits to avoid the values < 0 which are NaN data
	if min(datac[:, i-1]) >= 0 and min(datac[:, i]) >= 0:
		# Check if there are NaN values
		# If not :
		minx = min(np.min(datac[:, i-1]), np.min(datac[:, i]))
		maxx = max(np.nanmax(datac[:, i-1]), np.nanmax(datac[:, i]))
	else:
		# If yes :
		# find min and max without NaN data
		datanan = copy.copy(datac)
		datanan[datanan[:, i-1] < 0.] = 'NaN'
		datanan[datanan[:, i] < 0.] = 'NaN'
		minx = min(np.nanmin(datanan[:, i-1]), np.nanmin(datanan[:, i]))
		maxx = max(np.nanmax(datanan[:, i-1]), np.nanmax(datanan[:, i]))
	
	# plot the line 1:1 for comparison
	plt.plot([minx,maxx], [minx,maxx], '-', color = 'b')
	plt.xlim(xmin = minx, xmax = maxx)
	plt.ylim(ymin = minx, ymax = maxx)
	plt.xlabel(xlabel)
	plt.ylabel(ylabel)
	plt.grid(True)
	# Simplify the ticks (if not this is hard to read...)
	if i == 3:
		if (maxx-minx)/400 < 1:
			plt.xticks(np.arange(np.round(minx/10) * 10, 
			                    (np.round(maxx/10) +1) * 10, 
			                    np.round((maxx-minx)/40) * 10), 
			           rotation = 45)
			plt.yticks(np.arange(np.round(minx/10) * 10, 
			                     (np.round(maxx/10) +1) * 10, 
			                     np.round((maxx-minx)/40) * 10))
		elif (maxx-minx)/4000 < 1:
			plt.xticks(np.arange(np.round(minx/100) * 100, 
			           (np.round(maxx/100) +1) * 100, 
			            np.round((maxx-minx)/400) * 100), 
			            rotation = 45)
			plt.yticks(np.arange(np.round(minx/100) * 100, 
			                     (np.round(maxx/100) +1) * 100, 
			                     np.round((maxx-minx)/400) * 100))
		elif (maxx-minx)/40000 < 1:
			plt.xticks(np.arange(np.round(minx/1000) * 1000, 
			                     (np.round(maxx/1000) +1) * 1000, 
			                      np.round((maxx-minx)/4000) * 1000), 
			                      rotation = 45)
			plt.yticks(np.arange(np.round(minx/1000) * 1000, 
			                     (np.round(maxx/1000) +1) * 1000, 
			                      np.round((maxx-minx)/4000) * 1000))	
	else:
		if maxx-minx < 4:
			inter = 1
			if maxx-minx < 2:
				inter = 0.5
		else:
			inter = int((maxx-minx)/4)
		plt.xticks(np.arange(int(minx), int(maxx) + 1, inter))
		plt.yticks(np.arange(int(minx), int(maxx) + 1, inter))
	
	return
	# End of the function


############################################################################

def plotmtl(datac, iii, system, agecol, pgraph, pfolder):
	"""
	Function to plot the MTL data

	INPUT:
		datac   : 
		iii     : 
		system  : 
		agecol  : 
		pgraph  : 
		pfolder : 

	OUTPUTS:
		No outputs

	USAGE:
		plotmtl(datac, iii, system, agecol, pgraph, pfolder)
		
	"""
	
	# open a new figure
	#plt.figure(iii+3)
	lrange = np.arange(0.5,17.5)
	mtl = datac[iii,agecol[system]:agecol[system]+17]
	plt.bar(lrange, mtl, 
			widtch = 1, 
			bottom = None, 
			label='MTL',
			color = 'r', 
			alpha=0.5)
	mtlp = datac[iii,agecol[system]+18:agecol[system]+18+17]
	plt.bar(lrange, mtlp, 
			widtch = 1, 
			bottom = None, 
			label='Predicted MTL',
			color = 'b', 
			alpha=0.5)
	plt.xlable(agecol[system])
	plt.ylabel('Frequency')
	plt.legend(loc='best', numpoints = 1)
	plt.set_title('Sample number ' + str(iii + 1))
	
	return
	# End of the function

############################################################################

def find_nbplot(nbplot):
	"""
	Function to find the number line/columns of subplot 
	      depending on the total number of plots
	      
	      Written by Xavier Robert 2015.06.11
	      
	INPUT:
	   nbplot = total number of plots   

	OUTPUT:
	    nbplotx = nb of lines in the subplot
	    nbploty = nb of columns in the subplots

	USAGE:
	  nbplotx, nbploty = find_nbplot(nbplot)

	"""
	
	if nbplot == 1:
		nbplotx = 1
		nbploty = 1
	elif nbplot == 2:
		nbplotx = 2
		nbploty = 1
	elif nbplot in [3, 4]:
		nbplotx = 2
		nbploty = 2
	elif nbplot in [5, 6]:
		nbplotx = 3
		nbploty = 2
	elif nbplot in [7, 9]:
		nbplotx = 3
		nbploty = 3		
	elif nbplot in [10, 12]:
		nbplotx = 4
		nbploty = 3
	else:
		# If there more plots to do, it has not be yet implemented
		# Raise an error and give the clues to solve it !
		try:
			raise Exception("foo")
		except:
			for frame in traceback.extract_tb(sys.exc_info()[2]):
				fname,lineno,fn,text = frame
				sys.exit('ERROR: The number of parameters is more than 12 (in fatc %i). \n'
			         'You need to change the code before line %d' % (nbplot, lineno))
			         
	return nbplotx, nbploty


######## END of the functions definition ###########
############################################################################


############################################################################
######### Main code ###############
if __name__ == "__main__":
	#### Read the data
	#print(' ')
	print('___________________________________________\n')
	print('Plotting data for a Pecube forward model...')
	print('___________________________________________\n')
	
	if VPecube == 1:
		raise ValueError('ERROR: Pecube v1 not implemented, please upgrade the code\n')
	elif VPecube in [2,3]:
		print('WARNING: You are using Pecube v2 or V3, be carefull, this is an old version, please upgrade the code\n')
		# data: path and Name of the file Comparison.txt produced by Pecube
		#       Usually you do not have to change it
		data = 'Comparison.txt'

		# Read in which folder we are working from topo_parameters.txt
		with open('input/topo_parameters.txt', "r") as f0r:
			inputd = []
			for line in f0r:
				if line[0] != '$' and line[0] != '#' and line[0] != ' ' \
		   		and len(line) != 1 and len(line) != 0 :
					inputd.append(line.strip())
	
		# get the working folder (first input in topo_parameters.txt)
		pfolder = inputd[0]
		# get the ages data file (3rd input from the end of the file topo_parameters.txt ; 
		#     -3 should be changed if the end of topo_parameters.txt is changed)
		datain = inputd[-3]
		pgraph = str(pfolder) + '/' + graphpath
		pdata = pfolder + '/' + data
	elif VPecube == 4:
		print('You are using Pecube v4...\n')
		# data: path and Name of the comparison file produced by Pecube version 4+
		#       Usually you do not have to change it
		data   = 'output/CompareAge.csv'
		dataTT = 'output/CompareTT.csv'
		pdata = data

		with open('input/Pecube.in', "r") as f0r:
			lines = f0r.readlines()
			for line in lines:
				if "data_folder" in line:
					datain = 'data/' + line.split()[-1] + '/A_files.csv'
					datainTH = 'data/' + line.split()[-1] + '/TH_files.csv'
		pgraph = 'output/' + graphpath	

	print('\tWorking in the folder: ' + str(pfolder))
	
	# Check if the Graphs/ folder exists, if not create it
	if os.path.exists(pgraph) == False:
		print('Output folder does not exist...')
		print('I am creating it...\n')
		os.mkdir(pgraph)
	else:
		print('\tFolder %s/ already exists, I will write in it and erase previous pdf file\n' %(str(pgraph)))
	
	# Read file Comparison.txt or CompareAGE.csv
	# Check if file Comparison.txt exists, if not, raise an error
	if os.path.isfile(pdata) == False and os.access(pdata, os.R_OK) == False :
		raise FileExistsError('ERROR : File %s does not exist' %(str(pdata)))
	print('\tReading data...')
	if VPecube in [2,3]:
		datac = np.loadtxt(pdata, skiprows = 1)
	elif VPecube == 4: 
		#datac = np.genfromtxt(pdata, delimiter = ',', skip_header = 1)
		datac = np.genfromtxt(pdata, delimiter = ',', names = True)
		# Build a dictionnary to find the number of columns



		# PB : Names in header in CompareAGE.csv are not the same than in A_Files.csv (inputs)
		for item in tagsName:
			if item[0] == 'D':
				errcol[tags[item]] = datac.index[item]
			else:
				agecol[tags[item]] = datac.index[item]

	if datain != 'Nil':
		if os.path.isfile('Data/'+ datain) == False and os.access('Data/'+ datain, os.R_OK) == False :
			raise FileExistsError('ERROR : File Data/%s does not exist' %(str(datain)))
		if VPecube in [2,3]:
			dataerr = np.loadtxt('Data/'+ datain, skiprows = 1)
		elif VPecube == 4: 
			#dataerr = np.genfromtxt(datain, delimiter = ',', skip_header = 1)
			dataerr = np.genfromtxt(datain, delimiter = ',', names = True)
		dataerr[dataerr[:,:] < 0.] = np.nan
	else:
		dataerr[0:datac.shape[0], 0:datac.shape[1]]	= np.nan
	



	if datain != 'Nil':
		print('\tPlotting data')
		# Determine the number of plot
		nbplot = int(len(dataplot)) + 1
		nbplotx, nbploty = find_nbplot(nbplot)
		# Open the new figure
		plt.figure(1)
		# Adjust the subplots
		plt.subplots_adjust(wspace = (nbplot + 2)/10., hspace = (nbplot + 2)/10.)
		
		#### Plot the Altitude comparison
		plt.subplot(nbploty, nbplotx, 1, aspect='equal')
		plottot(datac, agecol['alt'], 'Elevation (m)', 'Predicted elevation (m)', err = None)
		
		for system in dataplot:
			if system != 'FTL':
				print('\t\tPlotting ' + system + ' system')
				plt.subplot(nbploty, nbplotx, 
				            dataplot.index(system) + 2, 
							aspect='equal')
				# use the dictionary defined at the beginning of the code
				plottot(datac, 
				        agecol[system], 
						agename[system], 
						predname[system], 
						err = dataerr[:, errcol[system] - 1])												
		# Add main title
		plt.suptitle('Run ' + str(pfolder), size=16)
		plt.savefig(pgraph + '/Plot_forward' + pfolder +'.pdf')
		plt.close(1)
		
		if 'FLT' in dataplot: 
			# Normalized FTL distribution as an array of length 17. 
			# ftld(k) is the proportion of tracks with length between k-0.5 and k+0.5 µm
			# the sum of ftlf(k) for k = 1 to 17 must be equal to 1
			# needs to be done for each given data > 0
			plt.figure(2)
			# Find the number of data to plot
			i = 0
			for iii in range (0, nbplot-1):
				if datac[iii, agecol[system]] >= 0:
					i += 1
			nbplotx, nbploty = find_nbplot(nbplot)
	
			# and only then, define the subplots
			plt.subplots_adjust(wspace = (nbplot + 2)/10., hspace = (nbplot + 2)/10.)
			print('\tPlotting FTL distributions')
			for iii in range (0, nbplot-1):
				plt.subplot(nbploty, nbplotx, iii + 1, aspect='equal')
				# stepping over data
				if datac[iii, agecol[system]] >= 0:
					print('\t\tDoing data number ' + str(iii) + ' on ' + str(nbplot))
					# if there are FTL datan call plotmtl
					plotmtl(datac, iii, system, agecol, pgraph, pfolder)
				else:
					print('\t\tData number %i on %i without MTL data' % (iii + 1, nbplot))
			plt.suptitle('MTL distributions')
			plt.savefig(pgraph + '/MTL_' + pfolder +'.pdf')
			plt.close(2)
		print(' ')
	else:
		print('No input data, no data/prediction graphs\n')
	






	# Plot data and prediction along simple transects (Altitude, Longitude or Latitude)
	for plotplot in profiletype:
		# The next few line are used to plot ages along a transect to compare data and model predictions. 
		# It could be changed, depending on your model settings
		# plot ages in function of latitude
		print('\tPlotting age transects %s...\n' % plotplot)
		plt.figure()
		if plotplot != 'Altitude':
			for system in dataplot:
				if system != 'FTL':
					datanan = copy.copy(datac)
					if datain != 'Nil': 
						plt.errorbar(datac[:, plotprof[plotplot]], datac[:, agecol[system]-1],
						             yerr =  dataerr[:, errcol[system] - 1],
						             fmt = 'o', 
						             color = colores[system], 
						             label = agename[system])
						# remove predictions where there is no data
						datanan[datanan[:, agecol[system]-1] < 0.] = 'NaN'
						for i in range (0,datanan.shape[0]):
							if datanan[i, agecol[system] - 1] == 'NaN':
								datanan[i, agecol[system]] == 'NaN'
								
					plt.plot(datac[:, plotprof[plotplot]], datanan[:, agecol[system]], 
					               's', 
					               color = colores[system], 
					               label = predname[system], 
					               alpha = 0.5)
			
		else:
			for system in dataplot:
				if system != 'FTL':
					datanan = copy.copy(datac)
					if datain != 'Nil':
						plt.errorbar(datac[:, agecol[system] - 1], datac[:, plotprof[plotplot]],
						             xerr =  dataerr[:, errcol[system] - 1],
						             fmt = 'o', 
						             color = colores[system], 
						             label = agename[system])
						# remove predictions where there is no data
						datanan[datanan[:, agecol[system]-1] < 0.] = 'NaN'
						for i in range (0,datanan.shape[0]):
							if datanan[i, agecol[system] - 1] == 'nan':
								datanan[i, agecol[system]] == 'NaN'
								
					plt.errorbar(datac[:, agecol[system]], datanan[:, plotprof[plotplot] + 1],
					         yerr = np.abs(datac[:, agecol['alt'] - 1] - datac[:, agecol['alt']]),
					         fmt = 's', 
					         color = colores[system], 
					         label = predname[system], 
					         alpha = 0.5)

		plt.grid(True)
		plt.legend(loc='best', numpoints = 1)
		# Find the age range. For that, choose 0 for the minimum, 
		#      and the max age whatever is the system used for the max range
		maxage = max(max(datac[:, agecol['AHe']]), 
		             max(datac[:, agecol['AFT']]), 
		             max(datac[:, agecol['ZFT']]),
		             max(datac[:, agecol['ZHe']]), 
		             max(datac[:, agecol['KAr']]), 
		             max(datac[:, agecol['MAr']]), 
		             max(datac[:, agecol['BAr']]), 
		             max(datac[:, agecol['HbAr']]))
		if plotplot != 'Altitude':
			# If the range of ages is not good on the graph, it can be set manually here
			# plt.xlim(0, your maxage)
			plt.ylim(0, maxage)
			plt.xlabel('%s ($^o$N)' % plotplot)
			plt.ylabel('Age (Ma)')
		else:
			# If the range of ages is not good on the graph, it can be set manually here
			# plt.xlim(0, your maxage)
			plt.xlim(0, maxage)
			plt.ylim(0, max(datac[:, plotprof[plotplot]]))
			plt.ylabel('%s (m)' % plotplot)
			plt.xlabel('Age (Ma)')
		plt.title('%s transect' % plotplot)
		plt.savefig(pgraph + '/Plot_forward_transect' + pfolder + '_' + plotplot +'.pdf')
		plt.close()
	
	print('\tSaving files...\n')
	# save input files in the working folder
	if VPecube in [2,3]:
		print('\t\t--> copying input/topo_parameters.txt to ' + pgraph)
		shutil.copy2('input/topo_parameters.txt', pgraph)
		print('\t\t--> copying input/fault_parameters.txt to ' + pgraph)
		shutil.copy2('input/fault_parameters.txt', pgraph)
	
	print('__________________________________________\n\n')
	
#### END ####